Lib/test/test_ntpath.py

```diff
@@ -369,19 +377,18 @@ def test_realpath_symlink_loops(self):
         self.assertPathEqual(ntpath.realpath(ABSTFN + "1\\..\\"
                                              + ntpath.basename(ABSTFN) + "y"),
                              ABSTFN + "x")
```

Comment:
I'm a bit surprised about these `assertPathEqual` asserts passing - looks like `_getfinalpathname()` normalizes before it resolves symlinks, whereas `readlink()` does not.

Reply:
`nt._getfinalpathname()` opens a handle to a file/directory with `CreateFileW()` and calls `GetFinalPathNameByHandleW()`. The latter makes a few system calls to get the final opened path in the filesystem (e.g. "\\Windows\\explorer.exe") and the canonical DOS name of the volume device on which the filesystem is mounted (e.g. "\\Device\\HarddiskVolume2" -> "\\\\?\\C:") in order to return a canonical DOS path (e.g. "\\\\?\\C:\\Windows\\explorer.exe").

Opening a handle with `CreateFileW()` entails first getting a fully-qualified and normalized NT path, which, among other things, entails resolving ".." components naively in the path string. This does not take reparse points such as symlinks and mountpoints into account. The only time Windows parses ".." components in an opened path the way POSIX does is in the kernel when they're in the target path of a relative symlink.

`nt.readlink()` opens a handle to the file with the flag `FILE_FLAG_OPEN_REPARSE_POINT`. If the final path component is a reparse point, it opens it instead of traversing it. Then the reparse point is read with the filesystem control request, `FSCTL_GET_REPARSE_POINT`. System symlinks and mountpoints (`IO_REPARSE_TAG_SYMLINK` and `IO_REPARSE_TAG_MOUNT_POINT`) are the only supported name-surrogate reparse-point types, though `os.stat()` and `os.lstat()` handle all name-surrogate types as 'links'. Moreover, only symlinks get the `S_IFLNK` mode flag in a stat result, because they're the only ones we can create with `os.symlink()` to satisfy the usage `if os.path.islink(src): os.symlink(os.readlink(src), dst)`.

Reply:
I think I follow! What would it take to do a POSIX-style "normalize as we resolve", and would we want to? I guess we'd need to call `nt._getfinalpathname()` on each path component in turn (`C:\, `C:\Users`, `C:\Users\Barney` etc), which from my pretty basic Windows knowledge might be rather slow if that involves file handles.

Perhaps another option would be to add a note to the `os.path.realpath()` documentation explaining the different normalization/resolution order across Windows and posix

Or perhaps we just ignore it, as a longstanding and esoteric issue? :)

Reply:
I'd add a note. Dangling symlinks or junctions ought to be rare, and in any case if you're resolving them then you ought to be prepared for the fact that you're doing something that's likely unnecessary, undesirable, and unreliable.

("Unnecessary and undesirable" may seem a bit strong, but the whole point is that you don't need to resolve a path to do 99% of actions, because the OS will do it. So resolve() is a fairly advanced function that I think is allowed to be closer to the OS behaviour than strictly consistent. (It might be worth adding this kind of note to the docs...))

Reply:
> What would it take to do a POSIX-style "normalize as we resolve",
> and would we want to? I guess we'd need to call nt._getfinalpathname()
> on each path component in turn (C:\, C:\Users, C:\Users\Barney etc),
> which from my pretty basic Windows knowledge might be rather slow if
> that involves file handles.

You asked, so I decided to write up an outline of what implementing a POSIX-style `realpath()` might look like in Windows. At its core, it's similar to POSIX: lstat(), and, for a symlink, readlink() and recur. The equivalent calls in Windows are the following:

* `CreateFileW()` (open a handle)
* `GetFileInformationByHandleEx()`: `FileAttributeTagInfo`
* `DeviceIoControl()`: `FSCTL_GET_REPARSE_POINT`

A symlink has the reparse tag `IO_REPARSE_TAG_SYMLINK`.

Filesystem mountpoints (aka junctions, which are like Unix bind mountpoints) must be retained in the resolved path in order to correctly resolve relative symlinks such as "\\spam" (relative to the resolved device) and "..\\..\\spam". Anyway, this is consistent with the UNC case, since mountpoints on a remote server can never be resolved (i.e. a final UNC path never resolves mountpoints).

Here are some of the notable differences:

* `GetFullPathNameW()` must be called initially if it's not a "\\\\?\\" path.
* Substitute drives (e.g. created by "subst.exe") and mapped drives (e.g. created by "net.exe") must be resolved, respectively via `QueryDosDeviceW()` and `WNetGetUniversalNameW()`. These are object symlinks (i.e. symlinks in the object namespace) that include an arbitrary number of path components, which may include filesystem symlinks that need to be resolved. The system object manager reparses all DOS devices, including these drives, to the path of an actual Device object before the I/O manager's parse routine ever sees the path. Such drives need to be resolved whenever parsing starts or restarts at a drive, but the result can be cached in case multiple symlinks target the same drive. 
    * Substitute drives can target paths on other substitute drives, so `QueryDosDeviceW()` has to be called in a loop that accumulates the tail path components until it reaches a real device (i.e. a target path that doesn't begin with "\\??\\"). 
    * `WNetGetUniversalNameW()` has to be called after resolving all substitute drives to a real device. It resolves a mapped drive to a UNC path. The target path of the object symlink that implements a mapped drive is of the opaque form "\\Device\\\<redirector device name\>\\;\<something\>\\server\\share\\some\\filesystem\\path". It's an implementation detail of the UNC redirector and the system's Multiple UNC Provider (MUP) device, which the Multiple Provider Router (MPR) API encapsulates. For example, for the mapped drive path "Z:\\spam\\eggs", it returns a UNC path of the form "\\\\server\\share\\some\\filesystem\\path\\spam\\eggs".
* A join that tries to resolve ".." against the drive or share root path must fail, whereas this is ignored for the root path in POSIX.
* At the end, the resolved path should be tested to try to remove "\\\\?\\" if the source path didn't have this prefix. Call `GetFullPathNameW()` to check for a reserved name in the final component and `PathCchCanonicalizeEx()` to check for long-path support. (The latter calls the system runtime library function `RtlAreLongPathsEnabled`, but that's an undocumented implementation detail.)

`GetFinalPathNameByHandleW()` is not required. Optionally, it can be called for the last valid component if the caller wants a final path with all mountpoints resolved, i.e. add a `final_path=False` option. Of course, a final UNC path must retain mountpoints, so there's nothing we can do in that case. It's fine that this `realpath()` implementation would return mountpoints in Windows (as the current implementation also does for UNC paths). They are not symlinks, and this matches the behavior of POSIX.

I'd include a warning in the documentation that getting a final path in the non-strict case may be dysfunctional. The unresolved tail end of the path may become valid again if a server or device comes back online. If the unresolved part contains symlinks with relative targets such as "\\spam" and "..\\..\\spam", and the `realpath()` call resolved away mountpoints, the reminaing path may not resolve correctly against the final path. It definitely will not resolve correctly for a rooted target path such as "\\spam", since it will reparse to the root path of the wrong device.

Reply:
Thanks for the detailed explanation, that's fantastic.

Are you happy for me to log a bug and include your write-up? Even with the caveats/incompatibility you mention, your proposal still seems like a substantial improvement.

Reply:
I've logged a follow-up bug here so this doesn't get lost: https://bugs.python.org/issue43936